### PR TITLE
Update documentation regarding annotations

### LINKED ISSUE #122: Documentation for PHP 8 attribute support is missing
Related to #113 

## Tasks

* Create new page with description for attributes
* Extract description for annotations from Quick Start to own page

## Page Structure

* Introduction
* Installation requirements (doctrine/annotations)
* Basic usage with code example (form, builder)
* all annotations / attributes with separate code example

### PR DESCRIPTION
**Pull request by driehle**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

#122

### CODE REVIEW

**Comment by driehle** [comment]
@froschdesign @Slamdunk before I proceed, can we agree on the proposed structure?

File: docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Advanced use cases
+# Using Custom Elements with laminas-servicemanager
```
Comment (froschdesign) [inline docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md]: Unfortunately the headline does not match the content of the page because some more topics are covered then custom elements. The topics of the page should be split, but until this, please revert the changes.

(And please remember: renaming a page leads to a new URL and therefore a redirect must be added.)

File: mkdocs.yml [hunk 2]
```diff
@@ -5,6 +5,14 @@ nav:
...
     - Installation: installation.md
     - "Quick Start": quick-start.md
+    - "Form Creation":
+        - "Programmatic form ceration": form-creation/programmatic-form-creation.md
+        - "Creation via factory": form-creation/creation-via-factory.md
+        - "Factory-backed form extension": form-creation/factory-backed-form-extension.md
+        - "Using annotations": form-creation/using-annotations.md
+    - "Application Integration":
+        - "Usage in a laminas-mvc application": application-integration/usage-in-a-laminas-mvc-application.md
+        - "Using custom elements with laminas-servicemanager": application-integration/using-custom-elements-with-laminas-servicemanager.md
```
Comment (froschdesign) [inline mkdocs.yml]:
```suggestion
    - "Form Creation":
        - "Programmatic Form Creation": form-creation/programmatic-form-creation.md
        - "Creation via Factory": form-creation/creation-via-factory.md
        - "Factory-Backed Form Extension": form-creation/factory-backed-form-extension.md
        - "Using Annotations or PHP8 Attributes": form-creation/using-annotations.md
    - "Application Integration":
        - "Usage in a laminas-mvc application": application-integration/usage-in-a-laminas-mvc-application.md
```

File: docs/book/quick-start.md [hunk 3]
```diff
@@ -12,399 +12,42 @@ and calling the `isValid()` method. If you want to simplify your work even more,
...
+`\Laminas\Form\Form`. In contrast to the creation via factory this bears the advantage 
+that you can override certain functionality programmatically, if needed, while 
+elements can still be added using the same configuration approach. See the page on
+[factory-backed form extension](form-creation/factory-backed-form-extension.md) 
+for further details and examples.
 
-use Laminas\Captcha\AdapterInterface as CaptchaAdapter;
-use Laminas\Form\Element;
-use Laminas\Form\Form;
+### Using Annotations
```
Comment (froschdesign) [inline docs/book/quick-start.md]:
```suggestion
### Using Annotations or PHP8 Attributes
```

File: docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md (hunk 1 above)
Comment (driehle) [reply docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md]: I disagree partly, everything is this page is related to laminas-servicemanager. How about naming the page "Usage with laminas-servicemanager"?

Alternatively, what would you suggest on how to split the page? And how can I add a redirect?

**Comment by froschdesign** [comment]
@driehle 
> …can we agree on the proposed structure?

The structure looks good! 👍 
The "Advanced use cases" page should be covered in a separate pull request, the focus for this pull request is "Use of annotations or PHP8 attributes". We don't want to ask too much of you. 😄 

Thanks for your great help!

File: docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md (hunk 1 above)
Comment (froschdesign) [reply docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md]: > I disagree partly, everything is this page is related to laminas-servicemanager.

I referred to "Using Custom Elements". The topics "Short names" and "Handling dependencies" are not part of this.

> Alternatively, what would you suggest on how to split the page?

I like your suggestion about custom elements because this is an important topic that needs to be highlighted more.

> And how can I add a redirect?

At the moment only with an _unbeautiful_ script tag. See: https://github.com/laminas/documentation-theme/issues/59

Comment (driehle) [reply docs/book/application-integration/using-custom-elements-with-laminas-servicemanager.md]: "unbeautiful" is indeed a good word for that... 😆

File: docs/book/form-creation/using-annotations.md [hunk 4]
```diff
@@ -6,84 +6,101 @@ representation for it, and potentially a hydrator for mapping the form elements
...
+PHPDoc annotations with v3, which have been supported with v2 already. Using
+these is suitable for legacy projects or if you need support for PHP 7.
+
+To use PHPDoc annotations, include them in your class and/or property docblocks.
 Annotation names will be resolved according to the import statements in your
 class; as such, you can make them as long or as short as you want depending on
 what you import.
 
-> ### doctrine/common dependency
+> ### doctrine/annotations dependency
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
> ### Installation Requirements
```

File: docs/book/form-creation/using-annotations.md [hunk 5]
```diff
@@ -6,84 +6,101 @@ representation for it, and potentially a hydrator for mapping the form elements
...
-  of validator options for the constructor:
-  `@Validator({"name": "StringLength", "options": {"min":3, "max": 25}})`.
-  This annotation may be specified multiple times.
-- `ContinueIfEmpty`: indicate whether the element can be submitted when it
-  is empty. A boolean value is expected. If `@Required` is set to `false`, this
-  needs to be set to `true` to allow the field to be empty.
-
-To use annotations, include them in your class and/or property docblocks.
+Annotations allow us to solve this problem. With v3, laminas-form supports two
+different types of annotations: PHP8 attributes and traditional PHPDoc annotations.
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]: > PHPDoc

We use "DocBlock" or "docblock" in all documentations, the preferred type is "DocBlock" as is used in [phpDocumentor documentation](https://phpdoc.org).

File: docs/book/form-creation/using-annotations.md [hunk 6]
```diff
@@ -6,84 +6,101 @@ representation for it, and potentially a hydrator for mapping the form elements
...
 what you import.
 
-> ### doctrine/common dependency
+> ### doctrine/annotations dependency
 >
-> Form annotations require `doctrine\common`, which contains an annotation
-> parsing engine. Install it using composer:
+> PHPDoc annotations require `doctrine\annotations` as a peer dependency, which 
+> contains an  annotation parsing engine. You need to manually install it using 
+> composer:
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
> DocBlock annotations require [Doctrine's annotations parser `doctrine\annotations`](https://www.doctrine-project.org/projects/annotations.html) as a peer dependency, which 
> contains an annotation parsing engine. You need to manually install it using 
> Composer:
```

File: docs/book/form-creation/using-annotations.md [hunk 7]
```diff
@@ -131,3 +139,330 @@ $form    = $builder->createForm(User::class);
...
+
+
+## List of Supported Annotations
+
+### AllowEmpty
+
+Marks an input as allowing an empty value. This annotation does
+not require a value. 
+
+*This annotation is deprecated, please add a `NotEmpty` validator instead.*
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
> ### Deprecated
>
> This annotation is deprecated, please add a `NotEmpty` validator instead.

Marks an input as allowing an empty value. This annotation does not require a value. 
```

File: docs/book/form-creation/using-annotations.md [hunk 8]
```diff
@@ -131,3 +139,330 @@ $form    = $builder->createForm(User::class);
...
+protected $myProperty;
+```
+
+### ContinueIfEmpty
+
+Indicate whether the element can be submitted when it is empty. A boolean
+value is expected. If `@Required` is set to `false`, this needs to be set
+to `true` to allow the field to be empty.
+
+*This annotation is deprecated, please add a `NotEmpty` validator instead.*
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
> ### Deprecated
>
> This annotation is deprecated, please add a `NotEmpty` validator instead.

Indicate whether the element can be submitted when it is empty. A boolean
value is expected. If `@Required` is set to `false`, this needs to be set
to `true` to allow the field to be empty.
```

File: docs/book/form-creation/using-annotations.md [hunk 9]
```diff
@@ -131,3 +139,330 @@ $form    = $builder->createForm(User::class);
...
+
+Marks an input as allowing an empty value. This annotation does
+not require a value. 
+
+*This annotation is deprecated, please add a `NotEmpty` validator instead.*
+
+```php
+/**
+* @AllowEmpty
+*/
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
/**
 * @AllowEmpty
 */
```
Some more examples contain the same problem.

File: docs/book/form-creation/using-annotations.md [hunk 10]
```diff
@@ -0,0 +1,468 @@
+# Using Annotations or PHP8 Attributes
```
Comment (froschdesign) [inline docs/book/form-creation/using-annotations.md]:
```suggestion
# Using PHP8 Attributes or Annotations
```

**Review by froschdesign — CHANGES_REQUESTED** [review]
Some small changes are necessary.

Thanks in advance! 👍

**Review by froschdesign — APPROVED** [review]
Great work!

**Comment by froschdesign** [comment]
@driehle 
Thank you for your time and this contribution! 👍

**Comment by Slamdunk** [comment]
Amazing work @driehle thank you
